pkg/scheduler/schedule_one.go

```diff
@@ -283,6 +283,15 @@ func (sched *Scheduler) bindingCycle(
 		return status
 	}
 
+	// All other extension points after here aren't supposed to fail in a usual state.
+	// i.e., They could return error only when some unexpected temporal errors happen. (e.g., kube-apiserver failure)
+	// So, if Pods are failed at one of them and going to the scheduling queue, the scheduling queue always pushes them to activeQ/backoffQ, instead of the unschedulable pod pool.
+	// This is because such Pods won't be schedulable by any certain cluster events, and thus scheduling hints aren't useful to determine when to push those Pods to activeQ/backoffQ.
+	//
+	// We can call Done() here because Pods failed at all other extension points after here won't use the scheduling hints as written.
+	// By calling Done(), we can free the cluster events stored in the scheduling queue, which is worth for busy clusters.
+	sched.SchedulingQueue.Done(assumedPod.UID)
```

Comment:
If, for example, this Pod is failed at bind, Done() is called again when pushing this Pod to the scheduling queue (in AddUnschedulableIfNotPresent). But, that's fine because Done() is created to be safe to call many times.
https://github.com/sanposhiho/kubernetes/blob/02fa38a74aeb55e247bde3ee650fcd7b639a12ef/pkg/scheduler/internal/queue/scheduling_queue.go#L868

Reply:
I'm not sure if it's safe to call Done more than once.

Keep in mind that, because of status updates, a Pod could enter the scheduling cycle more than once, even if the scheduling cycle succeeded.

Reply:
> a Pod could enter the scheduling cycle more than once

Oh, right, so the scenario like https://github.com/kubernetes/kubernetes/issues/118226.
OK, let's change here not to call Done() twice.

Reply:
Yes, that could. I meant to have a temporal solution (like have a flag representing whether we've called Done() or not) though, after trying some, looks like it's going to be not a simple one.
Let's just wait for #119607 and merge this PR after that.

/hold